cmd/deploy.go

```diff
@@ -221,7 +223,29 @@ func autofillApimodel(dc *deployCmd) {
 			// TODO: consider caching the creds here so they persist between subsequent runs of 'deploy'
 			appName := dc.containerService.Properties.MasterProfile.DNSPrefix
 			appURL := fmt.Sprintf("https://%s/", appName)
-			applicationID, servicePrincipalObjectID, secret, err := dc.client.CreateApp(appName, appURL)
+			var replyURLs *[]string
+			var requiredResourceAccess *[]graphrbac.RequiredResourceAccess
+			if dc.containerService.Properties.OrchestratorProfile.OrchestratorType == api.OpenShift {
+				appName = fmt.Sprintf("%s.%s.cloudapp.azure.com", appName, dc.containerService.Properties.AzProfile.Location)
+				appURL = fmt.Sprintf("https://%s:8443/", appName)
+				replyURLs = to.StringSlicePtr([]string{fmt.Sprintf("https://%s:8443/oauth2callback/Azure%%20AD", appName)})
+				ra := []graphrbac.ResourceAccess{
+					{
+						// TODO: where is this UUID defined?
+						ID:   to.StringPtr("311a71cc-e848-46a1-bdf8-97ff7156d8e6"),
```

Comment:
ew, what's this?

Reply:
These grant the application service principal the necessary rights to carry out oauth against AAD. I haven't found an obvious explanation to exactly what these UUIDs mean or predefined consts in a library anywhere though.

Reply:
This corresponds with the User.Read scope.  I can't find where it's documented anywhere, I expected it here: https://msdn.microsoft.com/library/azure/ad/graph/howto/azure-ad-graph-api-permission-scopes 

We should create a constant for this at least.  Found [this](https://github.com/Azure/AzureStack-Tools/blob/master/Identity/GraphAPI/GraphAPI.psm1#L113-L120) on github.  Note it is slightly different for China cloud in that code.

```
AadPermissions = [HashTable]@{
                    AccessDirectoryAsSignedInUser      = "a42657d6-7f20-40e3-b6f0-cee03008a62a"
                    EnableSignOnAndReadUserProfiles    = "311a71cc-e848-46a1-bdf8-97ff7156d8e6"
                    ReadAllGroups                      = "6234d376-f627-4f0f-90e0-dff25c5211a3"
                    ReadAllUsersBasicProfile           = "cba73afc-7f69-4d86-8450-4978e04ecd1a"
                    ReadAllUsersFullProfile            = "c582532d-9d9e-43bd-a97c-2667a28ce295"
                    ReadDirectoryData                  = "5778995a-e1bf-45b8-affa-663a9f3f4d04"
                    ManageAppsThatThisAppCreatesOrOwns = "824c81eb-e3f8-4ee6-8f6d-de7f50d565b7"
                }
```

and [this](http://www.cloudidentity.com/blog/2015/09/01/azure-ad-permissions-summary-table/) reference.